c10/util/Logging.h

```diff
@@ -316,6 +316,92 @@ C10_API bool LogAPIUsageFakeReturn(const std::string& context);
 // Initializes the c10 logger.
 C10_API void initLogging();
 
+class C10_API Log {
+ public:
+  Log(const char* component_alias,
+      int64_t py_log_level,
+      const SourceLocation& source_location,
+      std::string msg);
+
+  Log(const char* component_alias,
+      int64_t py_log_level,
+      const SourceLocation& source_location,
+      const char* msg);
+
+  Log(const char* component_alias,
+      int64_t py_log_level,
+      const SourceLocation& source_location,
+      ::c10::detail::CompileTimeEmptyString msg);
+
+  const std::string& component_alias() const;
+  int64_t py_log_level() const;
+  const SourceLocation& source_location() const;
+  const std::string& msg() const;
+
+ private:
+  // The name of the component under which to emit the log message.
+  std::string component_alias_;
+
+  // The Python log level to emit the log as.
+  int64_t py_log_level_;
+
+  // Where the log happened.
+  SourceLocation source_location_;
+
+  // The actual log message.
+  std::string msg_;
+};
+
+// Issue a log with a given message. Dispatched to the current
+// log handler.
+void C10_API log(const Log& log);
+
+class C10_API LogHandler {
+ public:
+  virtual ~LogHandler() = default;
+
+  virtual void process(const Log& log);
+};
+
+namespace LogUtils {
+
+/// Sets the global log handler. This is not thread-safe, so it should
+/// generally be called once during initialization or while holding the GIL
+/// for programs that use python.
+/// User is responsible for keeping the LogHandler alive until
+/// it is not needed.
+C10_API void set_log_handler(LogHandler* handler) noexcept(true);
```

Comment:
This isn't going to work for multipy, wondering what the strategy there will be.

Reply:
I'll look into it. If this doesn't work in multipy, then neither does `set_warning_handler`, since I just copied from that

Reply:
Essentially, the problem is that when you propagate the logs from C++ to Python, there is not a unique best handler for it; ANY of the Python interpreters could potentially be responsible for handling the logs. I suppose you could anoint a "primary" log handler, but in general it seems... suspicious to propagate logs to Python in a multipy setting; it seems better to just rely on some sort of reasonable C++ logging setup in this case...